sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -660,83 +678,113 @@ cdef class DistanceMetric{{name_suffix}}:
-        # to materialize the indices repetition explicitly.
-
-        n_Y, _ = Y.shape
-        Y_data = Y.reshape(-1)
-        Y_indices = np.arange(size, dtype=SPARSE_INDEX_TYPE)
-        Y_indptr = np.arange(
-            start=0, stop=size * (n_Y + 1), step=size, dtype=SPARSE_INDEX_TYPE
-        )
+        return np.asarray(Darr)
 
-        Darr = np.empty((n_X, n_Y), dtype=DTYPE, order='C')
-        self.cdist_csr(
-            x1_data=X_data,
-            x1_indices=X_indices,
-            x1_indptr=X_indptr,
-            x2_data=Y_data,
-            x2_indices=Y_indices,
-            x2_indptr=Y_indptr,
-            size=size,
-            D=Darr,
-        )
-        return Darr
-
-    def _pairwise_dense_sparse(self, X, Y):
-        # Same remark as in _pairwise_sparse_dense. We could
-        # have implemented this method using _pairwise_dense_sparse,
-        # but this would have come with an extra copy to ensure
-        # c-contiguity of the result.
-        n_Y, size = Y.shape
-        Y_data = np.asarray(Y.data, dtype={{INPUT_DTYPE}})
-        Y_indices = np.asarray(Y.indices, dtype=SPARSE_INDEX_TYPE)
-        Y_indptr = np.asarray(Y.indptr, dtype=SPARSE_INDEX_TYPE)
-
-        n_X, _ = X.shape
-        X_data = X.reshape(-1)
-        X_indices = np.arange(size, dtype=SPARSE_INDEX_TYPE)
-        X_indptr = np.arange(
-            start=0, stop=size * (n_X + 1), step=size, dtype=SPARSE_INDEX_TYPE
-        )
+    def _pairwise_sparse_dense(self,
+        X: csr_matrix,
+        const {{INPUT_DTYPE_t}}[:, ::1] Y,
+    ):
+        cdef:
+            ITYPE_t n_X = X.shape[0]
+            ITYPE_t n_features = X.shape[1]
+            const {{INPUT_DTYPE_t}}[:] X_data = np.asarray(X.data, dtype={{INPUT_DTYPE}})
+            const SPARSE_INDEX_TYPE_t[:] X_indices = np.asarray(X.indices, dtype=SPARSE_INDEX_TYPE)
+            const SPARSE_INDEX_TYPE_t[:] X_indptr = np.asarray(X.indptr, dtype=SPARSE_INDEX_TYPE)
+
+            ITYPE_t n_Y = Y.shape[0]
+            const {{INPUT_DTYPE_t}}[:] Y_data = np.reshape(Y, -1)
+            const SPARSE_INDEX_TYPE_t[:] Y_indices = (
+                np.arange(n_features, dtype=SPARSE_INDEX_TYPE)
+            )
 
-        Darr = np.empty((n_X, n_Y), dtype=DTYPE, order='C')
-        self.cdist_csr(
-            x1_data=X_data,
-            x1_indices=X_indices,
-            x1_indptr=X_indptr,
-            x2_data=Y_data,
-            x2_indices=Y_indices,
-            x2_indptr=Y_indptr,
-            size=size,
-            D=Darr,
-        )
-        return Darr
+            DTYPE_t[:, ::1] Darr = np.empty((n_X, n_Y), dtype=DTYPE, order='C')
+
+            ITYPE_t i1, i2
+            ITYPE_t x1_start, x1_end
+            {{INPUT_DTYPE_t}} * x2_data
+
+        with nogil:
+            # Use the exact same adaptation for CSR than in SparseDenseDatasetsPair
+            # for supporting the dense-sparse case with minimal overhead.
```

Comment:
I think we can anticipate #23958 in the inline comment.

```suggestion
            # for supporting the dense-sparse case with minimal overhead.
            # Note: at this point this method is only a convenience method
            # used in the tests via the DistanceMetric.pairwise method.
            # Therefore there is no need to attempt parallelization of those
            # nested for-loops.
            # Efficient parallel computation of pairwise distances can be
            # achieved via the PairwiseDistances class instead. The latter
            # internally calls into vector-wise distance computation from
            # the DistanceMetric subclass while benefiting from the generic
            # Cython/OpenMP parallelization template for the generic pairwise
            # distance + reduction computational pattern.
```

Reply:
Thank you for making this explicit.